src/librustc_resolve/lib.rs

```diff
@@ -797,36 +787,103 @@ bitflags! {
-    module_def: Option<Rc<Module>>,
-    type_def: Option<Def>,
-    type_span: Option<Span>
+    def: Option<Def>,
+    module: Option<Rc<Module>>,
+    span: Option<Span>
 }
 
-// Records a possibly-private value definition.
-#[derive(Clone, Copy, Debug)]
-struct ValueNsDef {
-    modifiers: DefModifiers, // see note in ImportResolution about how to use this
-    def: Def,
-    value_span: Option<Span>,
+impl NsDef {
+    fn def(&self) -> Option<Def> {
+        match (self.def, &self.module) {
+            (def @ Some(_), _) => def,
+            (_, &Some(ref module)) => module.def_id.get().map(|def_id| DefMod(def_id)),
+            _ => panic!("NsDef has neither a Def nor a Module"),
+        }
+    }
+}
+
+// Records at most one definition that a name in a namespace is bound to
+#[derive(Clone,Debug)]
+pub struct NameBinding(Rc<RefCell<Option<NsDef>>>);
+
+impl NameBinding {
+    fn new() -> Self {
+        NameBinding(Rc::new(RefCell::new(None)))
+    }
+
+    fn create_from_module(module: Rc<Module>) -> Self {
+        NameBinding(Rc::new(RefCell::new(Some(NsDef {
+            modifiers: DefModifiers::IMPORTABLE,
+            def: None,
+            module: Some(module),
+            span: None,
+        }))))
+    }
+
+    fn set(&self, modifiers: DefModifiers, def: Option<Def>, mod_: Option<Rc<Module>>, sp: Span) {
+        *self.0.borrow_mut() =
+            Some(NsDef { modifiers: modifiers, def: def, module: mod_, span: Some(sp) });
+    }
+
+    fn and_then<T, F: Fn(&NsDef) -> Option<T>>(&self, f: F) -> Option<T> {
+        self.borrow().as_ref().and_then(f)
+    }
+
+    fn borrow(&self) -> ::std::cell::Ref<Option<NsDef>> { self.0.borrow() }
+
+    // Lifted versions of the NsDef fields and method
+    fn def(&self) -> Option<Def>           { self.and_then(NsDef::def) }
+    fn span(&self) -> Option<Span>         { self.and_then(|def| def.span) }
+    fn module(&self) -> Option<Rc<Module>> { self.and_then(|def| def.module.clone()) }
+    fn modifiers(&self) -> Option<DefModifiers> { self.and_then(|def| Some(def.modifiers)) }
+
+    fn defined(&self) -> bool { self.borrow().is_some() }
+
+    fn defined_with(&self, modifiers: DefModifiers) -> bool {
+        self.modifiers().map(|m| m.contains(modifiers)).unwrap_or(false)
+    }
+
+    fn is_public(&self) -> bool {
+        self.defined_with(DefModifiers::PUBLIC)
+    }
+
+    fn def_and_lp(&self) -> (Def, LastPrivate) {
+        let def = self.def().unwrap();
+        (def, LastMod(if self.is_public() { AllPublic } else { DependsOn(def.def_id()) }))
+    }
 }
 
 // Records the definitions (at most one for each namespace) that a name is
 // bound to.
-#[derive(Debug)]
+#[derive(Clone,Debug)]
 pub struct NameBindings {
-    type_def: RefCell<Option<TypeNsDef>>,   //< Meaning in type namespace.
-    value_def: RefCell<Option<ValueNsDef>>, //< Meaning in value namespace.
+    type_ns: NameBinding,  //< Meaning in type namespace.
```

Comment:
I moved the `Rc` into the `NameBinding` so a `NameBinding` could be used in place of a `Rc<NameBindings>` when we only want a single name binding (i.e. a binding of a name in a given namespace), not a name binding for each namespace.

For example, a [`Target` struct](https://github.com/rust-lang/rust/blob/master/src/librustc_resolve/resolve_imports.rs#L87) is always used to represent a single name binding (and some extra information) but was defined with a `Rc<NameBindings>`. Once I moved `Rc` into `NameBinding` I was able to change it to use a `NameBinding` instead. Without moving the `Rc`, this would have been impossible without doing some major refactoring like putting the `NameBinding`s in an arena and adding lifetime parameters everywhere.